src/CMakeLists.txt

```diff
@@ -151,6 +141,64 @@ ELSE ()
 	ADD_FEATURE_INFO(cURL GIT_CURL "cURL for HTTP proxy support")
 ENDIF()
 
+# Auto-select TLS backend
+IF(USE_HTTPS STREQUAL ON)
+	IF (SECURITY_FOUND)
+		IF (SECURITY_HAS_SSLCREATECONTEXT)
+			SET(HTTPS_BACKEND "SecureTransport")
+		ELSE()
+			MESSAGE("-- Security framework is too old, falling back to OpenSSL")
+			SET(HTTPS_BACKEND "OpenSSL")
+		ENDIF()
+	ELSEIF(WINHTTP)
+		SET(HTTPS_BACKEND "WinHTTP")
+	ELSE()
+		SET(HTTPS_BACKEND "OpenSSL")
+	ENDIF()
+ELSE()
+	SET(HTTPS_BACKEND ${USE_HTTPS})
+ENDIF()
+
+IF(HTTPS_BACKEND STREQUAL "SecureTransport")
+	IF (NOT SECURITY_FOUND)
+		MESSAGE(FATAL_ERROR "Asked for SecureTransport TLS backend, but it wasn't found")
+	ENDIF()
+	IF (NOT SECURITY_HAS_SSLCREATECONTEXT)
+		MESSAGE(FATAL_ERROR "Asked for SecureTransport TLS backend, but it doesn't support SSLCreateContext")
+	ENDIF()
+
+	SET(GIT_SECURE_TRANSPORT 1)
+	LIST(APPEND LIBGIT2_INCLUDES ${SECURITY_INCLUDE_DIR})
+	LIST(APPEND LIBGIT2_LIBS ${SECURITY_LIBRARIES})
+	LIST(APPEND LIBGIT2_PC_LIBS ${SECURITY_LDFLAGS})
+ELSEIF(HTTPS_BACKEND STREQUAL "OpenSSL")
+	IF (NOT OPENSSL_FOUND)
+		MESSAGE(FATAL_ERROR "Asked for OpenSSL TLS backend, but it wasn't found")
+	ENDIF()
+
+	SET(GIT_OPENSSL 1)
+	LIST(APPEND LIBGIT2_INCLUDES ${OPENSSL_INCLUDE_DIR})
+	LIST(APPEND LIBGIT2_LIBS ${OPENSSL_LIBRARIES})
+	LIST(APPEND LIBGIT2_PC_LIBS ${OPENSSL_LDFLAGS})
+ELSEIF(HTTPS_BACKEND STREQUAL "WinHTTP")
+	# WinHTTP setup was handled in the WinHTTP-specific block above
+ELSEIF(NOT(HTTPS_BACKEND))
+	# The auto-selection wasn't able to auto-select, this will error below
+ELSE()
+	MESSAGE(FATAL_ERROR "Unknown backend ${HTTPS_BACKEND}")
+ENDIF()
+
+IF(USE_HTTPS AND NOT HTTPS_BACKEND)
+	MESSAGE(FATAL_ERROR "Asked for backend ${HTTPS_BACKEND} but it wasn't found")
```

Comment:
I think this cannot happen. If USE_HTTPS is set, then we'll always set it to a value with "OpenSSL" being the fallback. So in case of unknown backends we'll always fail with "OpenSSL not found" above, which is unfortunate.

Reply:
Actually, it should trigger when manually asking for a backend we don't support (eg. `-DUSE_HTTPS=gnuTLS`). Otherwise, with `-DUSE_HTTPS=ON` (the default) the automatic detection will auto-select OpenSSL, which will hard-fail if we can't find OpenSSL (which is what we want, since we can't obey USE_HTTPS).

I'm confused now... I'll try to come up with a cleaner set of `IF` blocks then.